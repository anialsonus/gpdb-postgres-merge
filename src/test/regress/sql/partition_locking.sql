-- Test locking behaviour. When creating, dropping, querying or adding indexes
-- partitioned tables, we want to lock only the master, not the children.

create or replace view locktest as
select coalesce(
  case when relname like 'pg_toast%index' then 'toast index'
       when relname like 'pg_toast%' then 'toast table'
       when relname like 'pg_aoseg%index' then 'aoseg index'
       when relname like 'pg_aoseg%' then 'aoseg table'
       else relname end, 'dropped table'),
  mode,
  locktype,
  l.gp_segment_id
from pg_locks l
left outer join pg_class c on (l.relation = c.oid),
pg_database d
where relation is not null
and l.database = d.oid
-- Show locks from master and one segment. We assume that all segments
-- will acquire the same locks, so we only need to check one of them.
-- Listing all of them would make the output differ depending on the
-- number of segments.
and l.gp_segment_id IN (-1, 0)
and (relname <> 'gp_fault_strategy' or relname IS NULL)
and d.datname = current_database()
order by 1, 3, 2;

-- Partitioned table with toast table
begin;

-- creation
create table g (i int, t text) partition by range(i)
(start(1) end(10) every(1));

select * from locktest;
commit;

-- drop
begin;
drop table g;
select * from locktest;
commit;

-- AO table (ao segments, block directory won't exist after create)
begin;
-- creation
create table g (i int, t text, n numeric)
with (appendonly = true)
partition by list(i)
(values(1), values(2), values(3));
select * from locktest;
commit;
begin;

-- add a little data
insert into g values(1), (2), (3);
insert into g values(1), (2), (3);
insert into g values(1), (2), (3);
insert into g values(1), (2), (3);
insert into g values(1), (2), (3);
select * from locktest;

commit;
-- drop
begin;
drop table g;
select * from locktest;
commit;

-- Indexing
create table g (i int, t text) partition by range(i)
(start(1) end(10) every(1));

begin;
create index g_idx on g(i);
select * from locktest;
commit;

-- Force use of the index in the select and delete below. We're not interested
-- in the plan we get, but a seqscan will not lock the index while an index
-- scan will, and we want to avoid the plan-dependent difference in the
-- expected output of this test.
set enable_seqscan=off;

-- test select locking
begin;
select * from g where i = 1;
-- Known_opt_diff: MPP-20936
select * from locktest;
commit;

begin;
-- insert locking
insert into g values(3, 'f');
select * from locktest;
commit;

-- delete locking
begin;
delete from g where i = 4;
-- Known_opt_diff: MPP-20936
select * from locktest;
commit;

-- drop index
begin;
drop table g;
select * from locktest;
commit;
