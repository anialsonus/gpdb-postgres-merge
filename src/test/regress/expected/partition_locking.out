-- Test locking behaviour. When creating, dropping, querying or adding indexes
-- partitioned tables, we want to lock only the master, not the children.
create or replace view locktest as
select coalesce(
  case when relname like 'pg_toast%index' then 'toast index'
       when relname like 'pg_toast%' then 'toast table'
       when relname like 'pg_aoseg%index' then 'aoseg index'
       when relname like 'pg_aoseg%' then 'aoseg table'
       else relname end, 'dropped table'),
  mode,
  locktype,
  l.gp_segment_id
from pg_locks l
left outer join pg_class c on (l.relation = c.oid),
pg_database d
where relation is not null
and l.database = d.oid
-- Show locks from master and one segment. We assume that all segments
-- will acquire the same locks, so we only need to check one of them.
-- Listing all of them would make the output differ depending on the
-- number of segments.
and l.gp_segment_id IN (-1, 0)
and (relname <> 'gp_fault_strategy' or relname IS NULL)
and d.datname = current_database()
order by 1, 3, 2;
-- Partitioned table with toast table
begin;
-- creation
create table g (i int, t text) partition by range(i)
(start(1) end(10) every(1));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "g_1_prt_1" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_2" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_3" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_4" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_5" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_6" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_7" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_8" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_9" for table "g"
select * from locktest;
          coalesce          |        mode         | locktype | gp_segment_id 
----------------------------+---------------------+----------+---------------
 g                          | AccessExclusiveLock | relation |             0
 g                          | AccessExclusiveLock | relation |            -1
 gp_distribution_policy     | RowExclusiveLock    | relation |            -1
 locktest                   | AccessShareLock     | relation |            -1
 pg_class                   | AccessShareLock     | relation |            -1
 pg_class                   | RowExclusiveLock    | relation |             0
 pg_class                   | RowExclusiveLock    | relation |            -1
 pg_class_oid_index         | AccessShareLock     | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock     | relation |            -1
 pg_depend                  | RowExclusiveLock    | relation |            -1
 pg_locks                   | AccessShareLock     | relation |            -1
 pg_partition               | AccessShareLock     | relation |            -1
 pg_partition               | RowExclusiveLock    | relation |            -1
 pg_partition_rule          | RowExclusiveLock    | relation |            -1
 pg_type                    | RowExclusiveLock    | relation |            -1
 toast index                | AccessExclusiveLock | relation |             0
 toast index                | AccessExclusiveLock | relation |            -1
 toast table                | ShareLock           | relation |             0
 toast table                | ShareLock           | relation |            -1
(19 rows)

commit;
-- drop
begin;
drop table g;
select * from locktest;
          coalesce          |        mode         | locktype | gp_segment_id 
----------------------------+---------------------+----------+---------------
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 gp_distribution_policy     | RowExclusiveLock    | relation |             0
 gp_distribution_policy     | RowExclusiveLock    | relation |            -1
 locktest                   | AccessShareLock     | relation |            -1
 pg_class                   | AccessShareLock     | relation |            -1
 pg_class                   | RowExclusiveLock    | relation |            -1
 pg_class_oid_index         | AccessShareLock     | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock     | relation |            -1
 pg_depend                  | RowExclusiveLock    | relation |            -1
 pg_locks                   | AccessShareLock     | relation |            -1
 pg_partition               | RowExclusiveLock    | relation |            -1
 pg_partition_rule          | RowExclusiveLock    | relation |            -1
 pg_type                    | RowExclusiveLock    | relation |            -1
(18 rows)

commit;
-- AO table (ao segments, block directory won't exist after create)
begin;
-- creation
create table g (i int, t text, n numeric)
with (appendonly = true)
partition by list(i)
(values(1), values(2), values(3));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "g_1_prt_1" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_2" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_3" for table "g"
select * from locktest;
          coalesce          |        mode         | locktype | gp_segment_id 
----------------------------+---------------------+----------+---------------
 g                          | AccessExclusiveLock | relation |             0
 g                          | AccessExclusiveLock | relation |            -1
 gp_distribution_policy     | RowExclusiveLock    | relation |            -1
 locktest                   | AccessShareLock     | relation |            -1
 pg_appendonly              | RowExclusiveLock    | relation |             0
 pg_appendonly              | RowExclusiveLock    | relation |            -1
 pg_class                   | AccessShareLock     | relation |            -1
 pg_class                   | RowExclusiveLock    | relation |             0
 pg_class                   | RowExclusiveLock    | relation |            -1
 pg_class_oid_index         | AccessShareLock     | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock     | relation |            -1
 pg_depend                  | RowExclusiveLock    | relation |            -1
 pg_locks                   | AccessShareLock     | relation |            -1
 pg_partition               | AccessShareLock     | relation |            -1
 pg_partition               | RowExclusiveLock    | relation |            -1
 pg_partition_rule          | RowExclusiveLock    | relation |            -1
 pg_type                    | RowExclusiveLock    | relation |            -1
 toast index                | AccessExclusiveLock | relation |             0
 toast index                | AccessExclusiveLock | relation |            -1
 toast table                | ShareLock           | relation |             0
 toast table                | ShareLock           | relation |            -1
(21 rows)

commit;
begin;
-- add a little data
insert into g values(1), (2), (3);
insert into g values(1), (2), (3);
insert into g values(1), (2), (3);
insert into g values(1), (2), (3);
insert into g values(1), (2), (3);
select * from locktest;
          coalesce          |        mode         |         locktype         | gp_segment_id 
----------------------------+---------------------+--------------------------+---------------
 aoseg index                | AccessShareLock     | relation                 |             0
 aoseg table                | AccessShareLock     | relation                 |             0
 aoseg table                | AccessShareLock     | relation                 |             0
 aoseg table                | AccessShareLock     | relation                 |             0
 aoseg table                | AccessShareLock     | relation                 |             0
 g                          | RowExclusiveLock    | relation                 |             0
 g                          | RowExclusiveLock    | relation                 |            -1
 g_1_prt_1                  | AccessExclusiveLock | append-only segment file |             0
 g_1_prt_1                  | AccessExclusiveLock | append-only segment file |            -1
 g_1_prt_1                  | AccessShareLock     | relation                 |            -1
 g_1_prt_1                  | RowExclusiveLock    | relation                 |             0
 g_1_prt_2                  | AccessExclusiveLock | append-only segment file |            -1
 g_1_prt_2                  | AccessShareLock     | relation                 |            -1
 g_1_prt_3                  | AccessExclusiveLock | append-only segment file |             0
 g_1_prt_3                  | AccessExclusiveLock | append-only segment file |            -1
 g_1_prt_3                  | AccessShareLock     | relation                 |            -1
 g_1_prt_3                  | RowExclusiveLock    | relation                 |             0
 locktest                   | AccessShareLock     | relation                 |            -1
 pg_class                   | AccessShareLock     | relation                 |            -1
 pg_class_oid_index         | AccessShareLock     | relation                 |            -1
 pg_class_relname_nsp_index | AccessShareLock     | relation                 |            -1
 pg_locks                   | AccessShareLock     | relation                 |            -1
(22 rows)

commit;
-- drop
begin;
drop table g;
select * from locktest;
          coalesce          |        mode         | locktype | gp_segment_id 
----------------------------+---------------------+----------+---------------
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 gp_distribution_policy     | RowExclusiveLock    | relation |             0
 gp_distribution_policy     | RowExclusiveLock    | relation |            -1
 locktest                   | AccessShareLock     | relation |            -1
 pg_appendonly              | RowExclusiveLock    | relation |             0
 pg_appendonly              | RowExclusiveLock    | relation |            -1
 pg_class                   | AccessShareLock     | relation |            -1
 pg_class                   | RowExclusiveLock    | relation |            -1
 pg_class_oid_index         | AccessShareLock     | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock     | relation |            -1
 pg_depend                  | RowExclusiveLock    | relation |            -1
 pg_locks                   | AccessShareLock     | relation |            -1
 pg_partition               | RowExclusiveLock    | relation |            -1
 pg_partition_rule          | RowExclusiveLock    | relation |            -1
 pg_type                    | RowExclusiveLock    | relation |            -1
(28 rows)

commit;
-- Indexing
create table g (i int, t text) partition by range(i)
(start(1) end(10) every(1));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "g_1_prt_1" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_2" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_3" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_4" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_5" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_6" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_7" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_8" for table "g"
NOTICE:  CREATE TABLE will create partition "g_1_prt_9" for table "g"
begin;
create index g_idx on g(i);
NOTICE:  building index for child partition "g_1_prt_1"
NOTICE:  building index for child partition "g_1_prt_2"
NOTICE:  building index for child partition "g_1_prt_3"
NOTICE:  building index for child partition "g_1_prt_4"
NOTICE:  building index for child partition "g_1_prt_5"
NOTICE:  building index for child partition "g_1_prt_6"
NOTICE:  building index for child partition "g_1_prt_7"
NOTICE:  building index for child partition "g_1_prt_8"
NOTICE:  building index for child partition "g_1_prt_9"
select * from locktest;
          coalesce          |        mode         | locktype | gp_segment_id 
----------------------------+---------------------+----------+---------------
 g                          | ShareLock           | relation |             0
 g                          | ShareLock           | relation |            -1
 g_idx                      | AccessExclusiveLock | relation |             0
 g_idx                      | AccessExclusiveLock | relation |            -1
 locktest                   | AccessShareLock     | relation |            -1
 pg_class                   | AccessShareLock     | relation |            -1
 pg_class                   | RowExclusiveLock    | relation |            -1
 pg_class_oid_index         | AccessShareLock     | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock     | relation |            -1
 pg_locks                   | AccessShareLock     | relation |            -1
 pg_partition               | AccessShareLock     | relation |            -1
 pg_partition_rule          | AccessShareLock     | relation |            -1
 pg_type                    | RowExclusiveLock    | relation |            -1
(13 rows)

commit;
-- Force use of the index in the select and delete below. We're not interested
-- in the plan we get, but a seqscan will not lock the index while an index
-- scan will, and we want to avoid the plan-dependent difference in the
-- expected output of this test.
set enable_seqscan=off;
-- test select locking
begin;
select * from g where i = 1;
 i | t 
---+---
(0 rows)

-- Known_opt_diff: MPP-20936
select * from locktest;
          coalesce          |      mode       | locktype | gp_segment_id 
----------------------------+-----------------+----------+---------------
 g                          | AccessShareLock | relation |            -1
 g_1_prt_1                  | AccessShareLock | relation |             0
 g_1_prt_1                  | AccessShareLock | relation |            -1
 g_idx_1_prt_1              | AccessShareLock | relation |             0
 g_idx_1_prt_1              | AccessShareLock | relation |            -1
 locktest                   | AccessShareLock | relation |            -1
 pg_class                   | AccessShareLock | relation |            -1
 pg_class_oid_index         | AccessShareLock | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock | relation |            -1
 pg_locks                   | AccessShareLock | relation |            -1
(10 rows)

commit;
begin;
-- insert locking
insert into g values(3, 'f');
select * from locktest;
          coalesce          |       mode       | locktype | gp_segment_id 
----------------------------+------------------+----------+---------------
 g                          | RowExclusiveLock | relation |             0
 g                          | RowExclusiveLock | relation |            -1
 g_1_prt_3                  | RowExclusiveLock | relation |             0
 locktest                   | AccessShareLock  | relation |            -1
 pg_class                   | AccessShareLock  | relation |            -1
 pg_class_oid_index         | AccessShareLock  | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock  | relation |            -1
 pg_locks                   | AccessShareLock  | relation |            -1
(8 rows)

commit;
-- delete locking
begin;
delete from g where i = 4;
-- Known_opt_diff: MPP-20936
select * from locktest;
          coalesce          |      mode       | locktype | gp_segment_id 
----------------------------+-----------------+----------+---------------
 g                          | ExclusiveLock   | relation |            -1
 g_1_prt_4                  | ExclusiveLock   | relation |            -1
 locktest                   | AccessShareLock | relation |            -1
 pg_class                   | AccessShareLock | relation |            -1
 pg_class_oid_index         | AccessShareLock | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock | relation |            -1
 pg_locks                   | AccessShareLock | relation |            -1
(7 rows)

commit;
-- drop index
begin;
drop table g;
select * from locktest;
          coalesce          |        mode         | locktype | gp_segment_id 
----------------------------+---------------------+----------+---------------
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |             0
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 dropped table              | AccessExclusiveLock | relation |            -1
 gp_distribution_policy     | RowExclusiveLock    | relation |             0
 gp_distribution_policy     | RowExclusiveLock    | relation |            -1
 locktest                   | AccessShareLock     | relation |            -1
 pg_class                   | AccessShareLock     | relation |            -1
 pg_class                   | RowExclusiveLock    | relation |            -1
 pg_class_oid_index         | AccessShareLock     | relation |            -1
 pg_class_relname_nsp_index | AccessShareLock     | relation |            -1
 pg_depend                  | RowExclusiveLock    | relation |            -1
 pg_locks                   | AccessShareLock     | relation |            -1
 pg_partition               | RowExclusiveLock    | relation |            -1
 pg_partition_rule          | RowExclusiveLock    | relation |            -1
 pg_type                    | RowExclusiveLock    | relation |            -1
(20 rows)

commit;
